pyevmasm/evmasm.py

```diff
@@ -1090,6 +596,293 @@ def assemble_hex(asmcode, pc=0, fork=DEFAULT_FORK):
         return '0x' + hexlify(b''.join([x.bytes for x in asmcode])).decode('ascii')
     return '0x' + hexlify(assemble(asmcode, pc=pc, fork=fork)).decode('ascii')
 
+
+class InstructionTable():
+    """
+    EVM Instruction factory
+        Implements an immutable, iterable instruction LUT that can be indexed by both mnemonic or opcode.
+
+    Example::
+
+        >>> from pyevmasm import instruction_tables
+        >>> instruction_table = instruction_tables['byzantium']
+        >>> instruction_table[0]
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> instruction_table['STOP']
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i = instruction_table.__iter__()
+        >>> i.__next__()
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i.__next__()
+        Instruction(0x1, 'ADD', 0, 2, 1, 3, 'Addition operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x2, 'MUL', 0, 2, 1, 5, 'Multiplication operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x3, 'SUB', 0, 2, 1, 3, 'Subtraction operation.', None, 0)
+
+    """
+    __slots__ = ('_previous_fork', '_ilist', '_cache', '_name_to_opcode')
+    def __init__(self, *args, **kwargs):
+        self._previous_fork = kwargs.get('previous_fork', None)
+        self._ilist = dict(args[0])
+        self._cache = {}
+        self._name_to_opcode = None
+
+    def _search_by_opcode(self, k):
+        try:
+            return (k, *self._ilist[k])
+        except KeyError:
+            if self._previous_fork is not None:
+                return self._previous_fork._search_by_opcode(k)
+        raise KeyError(k)
+
+    def _search_by_name(self, k):
+        if self._name_to_opcode is None:
+            self._name_to_opcode = {}
+            for opcode, (name, operand_size, pushes, pops, gas, description) in self._ilist.items():
+                if name == 'PUSH':
+                    long_name = 'PUSH%d' % operand_size
+                elif name == 'DUP':
+                    long_name = 'DUP%d' % pops
+                elif name == 'SWAP':
+                    long_name = 'SWAP%d' % (pops - 1)
+                elif name == 'LOG':
+                    long_name ='LOG%d' % (pops - 2)
+                else:
+                    long_name = name
+                self._name_to_opcode[long_name] = opcode
+
+        try:
+            return self._search_by_opcode(self._name_to_opcode[k])
+        except KeyError:
+            if self._previous_fork is not None:
+                return self._previous_fork._search_by_name(k)
```

Comment:
Hmm...
I think it would be better to create just two maps in `__init__` that would store all instructions for current fork (ofc create it based on current fork changes and previous forks). So we won't need to have `_previous_fork` at all and the search/fetch logic would be much easier?

Reply:
Ok...